### PR TITLE
pipes - Tamira - Media Ranker 

### PR DESCRIPTION
**Pull request by tvojnar**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a method in the Work model to sort all the Works by vote count in descending order. |
Describe how you approached testing that model method. What edge cases did you come up with? | One edge case I tested was what work would be returned as the media spotlight if there was a tie. |
Describe an edge case test you wrote for a controller | One edge case I tested was that the show method would return the status code not_found when the work being searched for does not exist |
What are `session` and `flash`? What is the difference between them? | Session and flash are special types of hashes that stick around between request cycles (they are sent to the browser, and then the browser sends them back in the next request cycle). The difference between them is that flash only sticks around for the next request cycle, but session sticks around until the user's session is over, when they log out or close the browser.  |
Describe a controller filter you wrote. | I wrote a controller filter for my find_work_by_params method. This method returns an instance variable that is set equal to the Work with id equal to params[:id]. i used this controller method in the work controller actions for show, edit, update, and destroy. 
What was one thing that you gained more clarity on through this assignment? | Rails testing and how to use model methods to access different instance variables in the views |
What is the Heroku URL of your deployed application | https://tv-media-ranker.herokuapp.com/|
Do you have any recommendations on how we could improve this project for the next cohort? | Have more in class project time to work on concepts we learn on the same day that we learn them. Also making it more clear that we don't need to test the session functionality. |

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  validates :name, presence: true
+end
+
+# QUESTION: I tried stripping white space from the beginning and end of the User name that was entered, but either it passed my tests but didn't work on my rails app, or it broke the test that a vote could be created with all fields. below are two examples of what I tried:
+
+  # 1.
+    # before_validation :strip_blanks
+    # def strip_blanks
+    #   self.name = self.name.strip!
+    # end
```
Comment (droberts-sea) [inline app/models/user.rb:14]: You were really close here! A model filter is definitely the way to go, the disconnect is in the return value of `String#strip!`.

`String#strip` (no `!`) creates a new string and returns it. `String#strip!` modifies the original string, but unlike `strip` it returns `nil` if the string wasn't modified.

So, you should either call `strip` and set `self.name` to the result, or call `strip!`, but not both. In other words:

```ruby
self.name = self.name.strip
# - OR -
self.name.strip!
```

[Here is the ruby doc describing this behavior](https://ruby-doc.org/core-2.4.2/String.html#method-i-strip-21) which, as far as I'm concerned, is really unintuitive. You're not the first student to get bitten by this.

Comment (droberts-sea) [reply app/models/user.rb:14]: I'm not familiar with `normaize_attribute`, is that from a separate gem?

File: test/controllers/works_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  describe "index" do
+    it "will return success status" do
+      get works_path
+      must_respond_with :success
+    end # it "will return success status" do
+  end # index
+
+  describe "root" do
+    # check that the index page works
+    it "returns success status" do
```
Comment (droberts-sea) [inline test/controllers/works_controller_test.rb:13]: For both index and root, I would be interested in what happens if there are no works, if there are works but no votes, and if one of the categories has no works.

File: test/controllers/works_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      # Assert
+      must_respond_with :redirect
+      must_redirect_to works_path
+
+      Work.count.must_equal start_count_work + 1
+    end # "it adds the work..."
+
+    it "returns bad_request when the work data is invalid" do
+      # QUESTION: what is invalid data? add this test if we add validations to the Work model
+    end # it "returns bad_request when the work data is invalid" do
```
Comment (droberts-sea) [inline test/controllers/works_controller_test.rb:54]: Here bad data would be anything that makes the `Work` validations fail, for example not having a title. The goal is not to test your specific validations, but to test that this action can handle bad data in general.

File: test/controllers/works_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+    it "returns not_found when given an invalid work id" do
+      invalid_work_id = Work.last.id + 1
+      start_work_count = Work.count
+
+      delete work_path(invalid_work_id)
+
+      must_respond_with :not_found
+      Work.count.must_equal start_work_count
+    end # returns invalid
```
Comment (droberts-sea) [inline test/controllers/works_controller_test.rb:150]: What happens if you try to delete a work with votes? Can you do it? What happens to those votes?

These are exactly the sort of question that the end-to-end style of controller tests are well-suited to answer.

File: test/models/work_test.rb [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+      it "will return the first work if no works have votes" do
+        # TODO: why does this test fail even though the titles match???
+        w4.save
+        w3.save
+
+        top = Work.top_work
+
+        top.title.must_equal w4.title
+      end # return when no votes
+
```
Comment (droberts-sea) [inline test/models/work_test.rb:121]: What if there are no works?

File: test/models/work_test.rb [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+      it "will return the top ten works of a category" do
+        ten_works = Work.top_ten("album")
+
+        length = ten_works.length
+        i = 0
+        while i < length
+          ten_works.votes[i].count.must_be :>=, ten_work.votes[i - 1].count
+          ten_works[i].category.must_equal 'album'
+        end
+      end # return the top ten works for a category
```
Comment (droberts-sea) [inline test/models/work_test.rb:135]: What if there are no works? What if there are more than 10 works? What if there are no votes, or if there's a tie between the 10th and 11th work?

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | yes
Comprehension questions	| yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | yes
Semantic HTML | yes - good use of partial views to keep HTML readable and DRY!
Errors are reported to the user | yes
Business logic lives in the models | yes - great work!
Models are thoroughly tested, including relations, validations and any custom logic | yes - There's some room to expand this coverage, but this is a good start. See inline comments.
Controllers are thoroughly tested | As with model testing, this is a good start.
**Wave 1 - Media** |
Splash page shows the three media categories | yes
Basic CRUD operations on media are present and functional | yes
**Wave 2 - Users and Votes** |
Users can log in and log out | I was not able to log in as a new user, but logging in as an existing user worked fine.
The ID of the current user is stored in the session | yes
Individual user pages and the user list are present | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes - link in the title of the work goes to the wrong page
Media pages contain lists of voting users | yes
**Wave 3 - Styling** |
Foundation is used appropriately | yes
Look and feel is similar to the original | yes - great work
**Overall** | Great job overall! This is a big project with many moving pieces, and I'm quite happy with this submission. The big place where I see room for improvement is around test coverage, but even this is a very good start. Keep up the hard work!
